src/Agoda.Analyzers/AgodaCustom/AG0018EnsureThatPubliclyExposedIEnumerableTypes .cs

```diff
@@ -0,0 +1,97 @@
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(CustomRulesResources.AG0018Title), CustomRulesResources.ResourceManager, typeof(CustomRulesResources));
+        private static readonly LocalizableString Description = DescriptionContentLoader.GetAnalyzerDescription(nameof(AG0018EnsureThatPubliclyExposedIEnumerableTypes));
+
+        private static readonly DiagnosticDescriptor Descriptor =
+          new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, AnalyzerCategory.CustomQualityRules,
+              DiagnosticSeverity.Warning, AnalyzerConstants.EnabledByDefault, Description, null, WellKnownDiagnosticTags.EditAndContinue);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Descriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeReturnType, SyntaxKind.MethodDeclaration);
+            context.RegisterSyntaxNodeAction(AnalyzePropertyType, SyntaxKind.PropertyDeclaration);
+        }
+
+        private void AnalyzeReturnType(SyntaxNodeAnalysisContext context)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+            if (IsPubliclyExposedIEnumerableTypes(context)) { return; }
+
+            context.ReportDiagnostic(Diagnostic.Create(Descriptor, methodDeclaration.GetLocation()));
+        }
+
+        private void AnalyzePropertyType(SyntaxNodeAnalysisContext context)
+        {
+            var propertyDeclarationSyntax = (PropertyDeclarationSyntax)context.Node;
+
+            if (IsPubliclyExposedIEnumerableTypes(context)) { return; }
+
+            context.ReportDiagnostic(Diagnostic.Create(Descriptor, propertyDeclarationSyntax.GetLocation()));
+        }
+
+        private bool IsPubliclyExposedIEnumerableTypes(SyntaxNodeAnalysisContext context)
+        {
+            INamedTypeSymbol namedTypeSymbol = null;
+            bool isArray = false;
+            if (context.ContainingSymbol.Kind == SymbolKind.Method)
+            {
+                IMethodSymbol method = (IMethodSymbol)context.ContainingSymbol;
+                isArray = method.ReturnType is IArrayTypeSymbol;
+                namedTypeSymbol = method.ReturnType as INamedTypeSymbol;
+                
+            }
+            else if (context.ContainingSymbol.Kind == SymbolKind.Property)
+            {
+                IPropertySymbol property = (IPropertySymbol)context.ContainingSymbol;
+                isArray = property is IArrayTypeSymbol;
+                namedTypeSymbol = property.Type as INamedTypeSymbol;
+                
+            }
+
+            if (isArray ||
+                    (namedTypeSymbol != null
+                     && namedTypeSymbol.ContainingNamespace.ToDisplayString().StartsWith("System.Collections")
```

Comment:
I'm not sure, but is this line even necessary? Seems like it would still work without it.

Reply:
The thing is f you have your own class which implemented IList<string> for the example. this case I'm not sure that we will allow them to return the implemented class or not. But in my opinion I thihnk it should be able to return the implemented one.